File: backend/core/utils/pseudocode_validator.py
"""
의사코드 로컬 검증 엔진
수정일: 2026-02-18
수정내용: 엔트로피 및 키워드 매칭 기반 무성의 입력(Low Effort) 감지 로직 추가 (프론트엔드 연동용)
"""
import re
import math
from collections import Counter

class PseudocodeValidator:
    def __init__(self, rules):
        self.rules = rules or self.get_default_rules()

    def validate(self, pseudocode):
        normalized = self.normalize(pseudocode)
        soft_normalized = self.soft_normalize(pseudocode)
        
        is_low_effort, low_effort_reason = self.check_low_effort(pseudocode)
        critical_errors = self.check_critical_errors(normalized)
        concepts = self.extract_concepts(soft_normalized)
        structure_results = self.analyze_structure(pseudocode, soft_normalized, concepts)
        
        return {
            'passed': len(critical_errors) == 0 and not is_low_effort,
            'score': 0 if is_low_effort else structure_results['score'],
            'is_low_effort': is_low_effort,
            'low_effort_reason': low_effort_reason,
            'criticalErrors': critical_errors,
            'warnings': structure_results['warnings'],
            'details': {
                'concepts': list(concepts),
                'structure': structure_results['feedback'],
                'flow': structure_results['flow']
            }
        }

    def check_low_effort(self, text: str):
        """성의 없는 입력인지 검증 (프론트엔드 동기화 및 강화)"""
        if not text or len(text.strip()) < 10:
            return True, "분석을 시작하기엔 설계 내용이 너무 짧습니다."
        
        # 1. 포기성/무의미 키워드
        giveup_keywords = [
            r'모르', r'몰라', r'몰겠', r'어렵', r'못하', r'안됨', r'해줘', r'\?', r'help',
            r'글쎄', r'나중에', r'다음에', r'귀찮', r'패스', r'pass', r'ㅁㄴㅇㄹ', r'ㄴㄴ'
        ]
        if any(re.search(kw, text) for kw in giveup_keywords):
            return True, "설계를 포기하시거나 질문을 하셨군요. 이 단계는 당신의 '설계안'을 제출하는 곳입니다."
            
        # 2. 엔트로피 검사 (무의미한 반복 문자열 aaaaa... 등)
        clean_text = "".join([c for c in text if c.isalnum()])
        if len(clean_text) > 5:
            counter = Counter(clean_text)
            probs = [count / len(clean_text) for count in counter.values()]
            entropy = -sum(p * math.log2(p) for p in probs)
            if entropy < 2.0:
                return True, "의미를 알 수 없는 단어의 나열이 감지되었습니다."
        
        return False, None
    def normalize(self, text):
        if not text: return ""
        text = text.lower()
        text = re.sub(r'\s+', ' ', text)
        # Keep essential characters for pseudocode
        text = re.sub(r'[^a-z0-9가-힣\s\.\,\(\)\_\-\:\;\=\>\<\!\?\/]', ' ', text)
        return text.strip()

    def soft_normalize(self, text):
        if not text: return ""
        return re.sub(r'\s+', ' ', text.lower()).strip()

    def check_critical_errors(self, normalized):
        errors = []
        critical_patterns = self.rules.get('criticalPatterns', [])
        
        for pattern_def in critical_patterns:
            if pattern_def.get('severity') in ['PRAISE', 'INFO']:
                continue
                
            pattern = pattern_def.get('pattern')
            is_error = False
            
            if isinstance(pattern, str): # Simple string or regex string
                if re.search(pattern, normalized, re.IGNORECASE):
                    is_error = True
            elif isinstance(pattern, dict):
                positive = pattern.get('positive')
                negatives = pattern.get('negatives', [])
                
                # In Python, we can pass regex objects or strings
                if positive:
                    if re.search(positive, normalized, re.IGNORECASE):
                        has_negative = any(re.search(neg, normalized, re.IGNORECASE) for neg in negatives)
                        if not has_negative:
                            is_error = True
            
            if is_error:
                errors.append({
                    'severity': pattern_def.get('severity', 'CRITICAL'),
                    'message': pattern_def.get('message'),
                    'example': pattern_def.get('correctExample'),
                    'why': pattern_def.get('explanation')
                })
        return errors

    def extract_concepts(self, soft_normalized):
        found_concepts = set()
        required_concepts = self.rules.get('requiredConcepts', [])
        
        for concept in required_concepts:
            patterns = concept.get('patterns', [])
            for pattern in patterns:
                if re.search(pattern, soft_normalized, re.IGNORECASE):
                    found_concepts.add(concept['id'])
                    break
        return found_concepts

    def analyze_structure(self, raw_pseudocode, soft_normalized, found_concepts):
        lines = [l.strip() for l in raw_pseudocode.split('\n') if l.strip()]
        scoring = self.rules.get('scoring', {'structure': 20, 'concepts': 40, 'flow': 40})
        
        feedback = []
        warnings = []
        score = 0
        
        # 1. Basic Structure (Lines, Numbering)
        rec = self.rules.get('recommendations', {})
        min_lines = rec.get('minLines', 3)
        max_lines = rec.get('maxLines', 20)
        
        if min_lines <= len(lines) <= max_lines:
            score += scoring['structure'] / 2
            feedback.append('✅ 적절한 길이')
        else:
            feedback.append(f"⚠️ {'너무 짧음' if len(lines) < min_lines else '너무 김'}")
            
        has_numbering = any(re.match(r'^\d+[\.\):]', l) for l in lines)
        if has_numbering:
            score += scoring['structure'] / 2
            feedback.append('✅ 번호 매기기 사용')
        
        # 2. Concepts
        required_concepts = self.rules.get('requiredConcepts', [])
        if required_concepts:
            total_weight = sum(c.get('weight', 1) for c in required_concepts)
            found_weight = sum(c.get('weight', 1) for c in required_concepts if c['id'] in found_concepts)
            
            concept_score = (found_weight / total_weight) * scoring['concepts'] if total_weight > 0 else scoring['concepts']
            score += concept_score
            
            missing = [c['name'] for c in required_concepts if c['id'] not in found_concepts]
            if not missing:
                feedback.append('✅ 모든 핵심 개념 포함')
            else:
                feedback.append(f"⚠️ 누락된 개념: {', '.join(missing)}")
                warnings.append(f"💡 누락된 개념을 추가해 보세요: {', '.join(missing)}")
        
        # 3. Flow
        flow_results = self.analyze_flow(soft_normalized, scoring['flow'])
        score += flow_results['score']
        feedback.extend(flow_results['feedback'])
        
        return {
            'score': round(score),
            'feedback': feedback,
            'warnings': warnings,
            'flow': flow_results
        }

    def analyze_flow(self, soft_normalized, max_score):
        lines = soft_normalized.split('\n')
        score = 0
        feedback = []
        
        dependencies = self.rules.get('dependencies', [])
        if not dependencies:
            return {'score': max_score, 'feedback': []}
            
        total_points = sum(d.get('points', 0) for d in dependencies)
        if total_points == 0:
            return {'score': max_score, 'feedback': []}
            
        for dep in dependencies:
            before_idx = self.find_concept_line(lines, dep['before'])
            after_idx = self.find_concept_line(lines, dep['after'])
            
            if before_idx != -1 and after_idx != -1:
                if before_idx < after_idx:
                    score += (dep.get('points', 0) / total_points) * max_score
                    feedback.append(f"✅ {dep.get('name', '순서')} 정확")
                else:
                    if dep.get('strictness') == 'REQUIRED':
                        feedback.append(f"❌ {dep.get('name', '순서')}: 순서 오류 (필수)")
                    else:
                        feedback.append(f"⚠️ {dep.get('name', '순서')}: 순서 권장됨")
                        score += ((dep.get('points', 0) / 2) / total_points) * max_score
                        
        return {'score': round(score), 'feedback': feedback}

    def find_concept_line(self, lines, concept_id):
        required_concepts = self.rules.get('requiredConcepts', [])
        concept = next((c for c in required_concepts if c['id'] == concept_id), None)
        if not concept: return -1
        
        patterns = concept.get('patterns', [])
        for i, line in enumerate(lines):
            for p in patterns:
                if re.search(p, line, re.IGNORECASE):
                    return i
        return -1

    def get_default_rules(self):
        return {
            'criticalPatterns': [],
            'requiredConcepts': [],
            'dependencies': [],
            'scoring': {'structure': 20, 'concepts': 40, 'flow': 40}
        }

File: frontend/src/features/practice/pseudocode/utils/PseudocodeValidator.js
/**
 * PseudocodeValidator.js - Rule 기반 검증 엔진 (클라이언트)
 * 수정일: 2026-02-19
 *
 * [변경 사항]
 * - isMeaningfulInput 제거 → LowEffortDetector.js로 이동
 * - 이 클래스는 Rule 기반 구조/개념/흐름 검증만 담당합니다.
 */

export class PseudocodeValidator {
    constructor(problem) {
        this.problem = problem;
        this.rules = problem?.validation || this._defaultRules();
    }

    // ── 메인 검증 ───────────────────────────────────────────────

    validate(pseudocode) {
        const criticalErrors = this._checkCriticalErrors(pseudocode);
        const structure = this._analyzeStructure(pseudocode);
        const warnings = this._generateWarnings(pseudocode, structure);

        return {
            passed: criticalErrors.length === 0,
            score: structure.score,
            criticalErrors,
            warnings,
            details: {
                structure,
                concepts: this._extractConcepts(pseudocode),
                completeness: this._checkCompleteness(pseudocode),
            },
        };
    }

    // ── 치명적 오류 체크 ─────────────────────────────────────────

    _checkCriticalErrors(pseudocode) {
        const errors = [];
        const normalized = this._normalize(pseudocode);
        if (!this.rules.criticalPatterns) return errors;

        for (const p of this.rules.criticalPatterns) {
            if (p.severity === 'PRAISE' || p.severity === 'INFO') continue;

            let isError = false;
            const { pattern } = p;

            if (pattern instanceof RegExp) {
                isError = pattern.test(normalized);
            } else if (pattern && typeof pattern === 'object') {
                const { positive, negatives = [] } = pattern;
                if (positive?.test?.(normalized)) {
                    const hasNeg = negatives.some((n) => n?.test?.(normalized));
                    if (!hasNeg) isError = true;
                }
            } else if (typeof pattern === 'function') {
                try { isError = pattern(normalized); } catch { isError = false; }
            }

            if (isError) {
                errors.push({
                    severity: p.severity || 'CRITICAL',
                    message: p.message,
                    example: p.correctExample,
                    why: p.explanation,
                });
            }
        }
        return errors;
    }

    // ── 구조 분석 ────────────────────────────────────────────────

    _analyzeStructure(pseudocode) {
        const lines = pseudocode.split('\n').filter((l) => l.trim());
        const scoring = this.rules.scoring || { structure: 20, concepts: 40, flow: 40 };

        let score = 0;
        const feedback = [];

        const structureScore = this._evalBasicStructure(lines, scoring.structure);
        score += structureScore.score;
        feedback.push(...structureScore.feedback);

        const concepts = this._extractConcepts(pseudocode);
        const conceptScore = this._evalConcepts(concepts, scoring.concepts);
        score += conceptScore.score;
        feedback.push(...conceptScore.feedback);

        const flow = this._analyzeFlow(pseudocode, concepts, scoring.flow);
        score += flow.score;
        feedback.push(...flow.feedback);

        return { score: Math.min(100, score), feedback, concepts: Array.from(concepts), flow };
    }

    _evalBasicStructure(lines, maxScore) {
        let score = 0;
        const feedback = [];
        const rec = this.rules.recommendations || {};
        const min = rec.minLines || 3;
        const max = rec.maxLines || 20;

        if (lines.length >= min && lines.length <= max) {
            score += maxScore / 2;
            feedback.push('✅ 적절한 길이');
        } else {
            feedback.push(`⚠️ ${lines.length < min ? '너무 짧음' : '너무 김'}`);
        }

        if (lines.some((l) => /^\d+[\.\):]/.test(l.trim()))) {
            score += maxScore / 2;
            feedback.push('✅ 번호 매기기 사용');
        }
        return { score, feedback };
    }

    _extractConcepts(pseudocode) {
        const normalized = this._softNormalize(pseudocode);
        const found = new Set();
        for (const concept of this.rules.requiredConcepts || []) {
            for (const pattern of concept.patterns || []) {
                try {
                    if (pattern?.test?.(normalized)) { found.add(concept.id); break; }
                } catch { /* skip */ }
            }
        }
        return found;
    }

    _evalConcepts(concepts, maxScore) {
        const required = this.rules.requiredConcepts || [];
        if (!required.length) return { score: maxScore, feedback: [] };

        const totalW = required.reduce((s, c) => s + (c.weight || 1), 0);
        const foundW = required.filter((c) => concepts.has(c.id)).reduce((s, c) => s + (c.weight || 1), 0);
        const score = totalW > 0 ? Math.round(maxScore * (foundW / totalW)) : 0;

        const missing = required.filter((c) => !concepts.has(c.id)).map((c) => c.name);
        const feedback = missing.length === 0
            ? ['✅ 모든 핵심 개념 포함']
            : [`⚠️ 누락된 개념: ${missing.join(', ')}`];

        return { score, feedback };
    }

    _analyzeFlow(pseudocode, concepts, maxScore) {
        const lines = this._softNormalize(pseudocode).split('\n');
        const deps = this.rules.dependencies || [];
        if (!deps.length) return { score: maxScore, feedback: [] };

        const totalPts = deps.reduce((s, d) => s + (d.points || 0), 0);
        if (!totalPts) return { score: maxScore, feedback: [] };

        let score = 0;
        const feedback = [];
        for (const dep of deps) {
            const bi = this._findConceptLine(lines, dep.before);
            const ai = this._findConceptLine(lines, dep.after);
            if (bi === -1 || ai === -1) continue;
            const pts = dep.points || 0;
            if (bi < ai) {
                score += (pts / totalPts) * maxScore;
                feedback.push(`✅ ${dep.name} 순서 정확`);
            } else if (dep.strictness === 'REQUIRED') {
                feedback.push(`❌ ${dep.name}: 순서 오류 (필수)`);
            } else {
                score += (pts / 2 / totalPts) * maxScore;
                feedback.push(`⚠️ ${dep.name}: 순서 권장됨`);
            }
        }
        return { score: Math.round(score), feedback };
    }

    _findConceptLine(lines, conceptId) {
        const concept = (this.rules.requiredConcepts || []).find((c) => c.id === conceptId);
        if (!concept) return -1;
        for (let i = 0; i < lines.length; i++) {
            for (const p of concept.patterns || []) {
                try { if (p?.test?.(lines[i])) return i; } catch { /* skip */ }
            }
        }
        return -1;
    }

    // ── 정규화 ────────────────────────────────────────────────────

    _normalize(text) {
        if (!text) return '';
        return text
            .toLowerCase()
            .replace(/\s+/g, ' ')
            .replace(/[^a-z0-9가-힣\s\.\,\(\)\_\-\:\;\=\>\<\!\?\/]/g, ' ')
            .trim();
    }

    _softNormalize(text) {
        if (!text) return '';
        return text.toLowerCase().replace(/\s+/g, ' ').trim();
    }

    // ── 완성도 / 경고 ─────────────────────────────────────────────

    _checkCompleteness(pseudocode) {
        const wordCount = pseudocode.split(/\s+/).filter((w) => w.length > 0).length;
        const rec = this.rules.recommendations || {};
        const minW = rec.minWords || 20;
        const maxW = rec.maxWords || 200;
        return {
            wordCount,
            adequate: wordCount >= minW && wordCount <= maxW,
            message: wordCount < minW
                ? `의사코드가 너무 간략합니다 (최소 ${minW}단어 권장)`
                : wordCount > maxW
                    ? `너무 세부적입니다 (최대 ${maxW}단어 권장)`
                    : '적절한 길이입니다.',
        };
    }

    _generateWarnings(pseudocode, structure) {
        const warnings = [];
        const completeness = this._checkCompleteness(pseudocode);
        if (!completeness.adequate) warnings.push(completeness.message);

        if (this.rules.recommendations?.exceptionHandling) {
            const soft = this._softNormalize(pseudocode);
            if (!/예외|오류|체크|검증|validation|error|check/i.test(soft)) {
                warnings.push('💡 예외 상황 처리를 추가하면 더 견고한 설계가 됩니다.');
            }
        }
        return warnings;
    }

    _defaultRules() {
        return {
            criticalPatterns: [],
            requiredConcepts: [
                { id: 'input',   name: '입력', weight: 1, patterns: [/입력|input|받|receive/i] },
                { id: 'process', name: '처리', weight: 1, patterns: [/처리|계산|process|compute/i] },
                { id: 'output',  name: '출력', weight: 1, patterns: [/출력|반환|return|output/i] },
            ],
            dependencies: [],
            scoring: { structure: 20, concepts: 40, flow: 40 },
            recommendations: { exceptionHandling: false, minLines: 3, maxLines: 20, minWords: 20, maxWords: 200 },
        };
    }
}

/**
 * 코드 검증 헬퍼 (주석 제거 후 검증)
 */
export class CodeValidator {
    constructor(rules) {
        this.rules = rules || {};
    }

    removeComments(code) {
        const patterns = this.rules.commentPatterns || [
            /#.*$/gm, /"""[\s\S]*?"""/g, /'''[\s\S]*?'''/g,
            /\/\/.*$/gm, /\/\*[\s\S]*?\*\//g,
        ];
        return patterns.reduce((c, p) => c.replace(p, ''), code);
    }

    validate(code) {
        const clean = this.removeComments(code);
        const errors = [];
        const warnings = [];

        for (const call of this.rules.requiredCalls || []) {
            if (!call.pattern.test(clean)) errors.push(`❌ ${call.name} 호출 누락`);
        }
        for (const f of this.rules.forbiddenPatterns || []) {
            const src = f.excludeComments ? clean : code;
            if (f.pattern.test(src)) errors.push(`🚨 ${f.message}`);
        }
        return { passed: errors.length === 0, errors, warnings, cleanCode: clean };
    }
}
